File: LeetCode/number_of_islands.cpp
class Solution {
public:
    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, -1, 1};
    int n;
    int m;
    vector< vector<bool> > seen;
    
    bool OK(vector<vector<char>>& grid, int x, int y) {
        if(x < 0 || x >= n)
            return false;
        
        if(y < 0 || y >= m)
            return false;
        
        if(seen[x][y] || grid[x][y] == '0')
            return false;
        
        return true;
    }
    
    void solve(vector<vector<char>>& grid, int x, int y) {
        seen[x][y] = true;
        
        for(int i = 0; i < 4; i++) {
            if(OK(grid, x + dx[i], y + dy[i]))
                solve(grid, x + dx[i], y + dy[i]);
        }
    }
    
    int numIslands(vector<vector<char>>& grid) {
        if(!grid.size())
            return 0;
        
        n = grid.size();
        m = grid[0].size();
        seen.assign(n, vector<bool> (m));
        
        int count = 0;
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(!seen[i][j] && grid[i][j] == '1') {
                    solve(grid, i, j);
                    count++;
                }
            }
        }
        
        return count;
    }
};

File: LeetCode/number_of_islands.go
var (
    dx = [4]int{1, -1, 0, 0}
    dy = [4]int{0, 0, 1, -1}
)

func canVisit(grid [][]byte, i, j int) bool {
    if i < 0 || i >= len(grid) {
        return false
    }

    if j < 0 || j >= len(grid[i]) {
        return false
    }

    return grid[i][j] == '1'
}

func dfs(grid [][]byte, i, j int, seen [][]bool) {
    seen[i][j] = true

    for k := 0; k < len(dx); k++ {
        x := i + dx[k]
        y := j + dy[k]

        if canVisit(grid, x, y) && !seen[x][y] {
            dfs(grid, x, y, seen)
        }
    }
}

func numIslands(grid [][]byte) int {
    seen := make([][]bool, len(grid)) 

    for i := range seen {
        seen[i] = make([]bool, len(grid[i]))
    }

    islands := 0
    for i := 0; i < len(grid); i++ {
        for j := 0; j < len(grid[i]); j++ {
            if grid[i][j] == '1' && !seen[i][j] {
                dfs(grid, i, j, seen)
                islands++
            }
        }
    }

    return islands
}
